File: frontend/src/lib/types/payment.ts
import { ID, ISODate } from "./common"

export type PaymentStatus =
    | "pending"
    | "success"
    | "failed"

export interface Payment {
    id: ID
    payment_method: "mpesa" | "card" | "cash"
    amount: number
    status: PaymentStatus
    order_reference: string
    transaction_id?: string | null
    mpesa_receipt?: string | null
    phone_number?: string | null
    failure_reason?: string | null
    created_at?: ISODate
    updated_at?: ISODate
}

File: backend/app/models/payment.py
"""
Payment model for M-Pesa and COD transactions
"""
from datetime import datetime

from app.extensions import db


class Payment(db.Model):
    """Payment transaction model"""
    __tablename__ = 'payments'

    id = db.Column(db.Integer, primary_key=True)
    order_reference = db.Column(db.String(50), nullable=False)
    amount = db.Column(db.Float, nullable=False)
    payment_method = db.Column(
        db.String(50), nullable=False)  # 'COD' or 'MPESA'
    status = db.Column(db.String(50), nullable=False, default='Pending')
    failure_reason = db.Column(db.Text, nullable=True)

    # M-Pesa specific fields
    transaction_id = db.Column(db.String(100), nullable=True)
    mpesa_receipt = db.Column(db.String(100), nullable=True)
    phone_number = db.Column(db.String(15), nullable=True)
    checkout_request_id = db.Column(db.String(100), nullable=True)
    merchant_request_id = db.Column(db.String(100), nullable=True)
    result_code = db.Column(db.String(10), nullable=True)
    result_desc = db.Column(db.Text, nullable=True)

    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    updated_at = db.Column(
        db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    def __repr__(self):
        return f'<Payment {self.order_reference} - {self.status}>'
